devices/uart-nrf9160/uart.c

```diff
@@ -0,0 +1,367 @@
+/*
+ * Phoenix-RTOS
+ *
+ * plo - operating system loader
+ *
+ * nRF9160 UART driver
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Damian Loewnau
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <hal/hal.h>
+#include <lib/errno.h>
+#include <lib/lib.h>
+#include <devices/devs.h>
+
+#include <board_config.h>
+
+
+#define BUFFER_SIZE 0x20
+
+
+typedef struct {
+	volatile unsigned int *base;
+	unsigned int cnt;
+	unsigned int irq;
+
+	volatile char *tx_dma;
+	volatile char *rx_dma;
+
+	u8 dataRx[BUFFER_SIZE];
+	cbuffer_t cbuffRx;
+
+	u8 dataTx[BUFFER_SIZE];
+	cbuffer_t cbuffTx;
+} uart_t;
+
+
+struct {
+	uart_t uarts[UART_MAX_CNT];
+} uart_common;
+
+/* clang-format off */
+enum { uarte_startrx = 0, uarte_stoprx, uarte_starttx, uarte_stoptx, uarte_flushrx = 11,
+	   uarte_events_cts = 64, uarte_events_rxdrdy = 66, uarte_events_endrx = 68, uarte_events_txdrdy = 71,
+	   uarte_events_endtx, uarte_events_error, uarte_events_rxto = 81, uarte_events_txstarted = 84,
+	   uarte_inten = 192, uarte_intenset, uarte_intenclr, uarte_errorsrc = 288, uarte_enable = 320,
+	   uarte_psel_rts = 322, uarte_psel_txd, uarte_psel_cts, uarte_psel_rxd, uarte_baudrate = 329,
+	   uarte_rxd_ptr = 333, uarte_rxd_maxcnt, uarte_rxd_amount, uarte_txd_ptr = 337, uarte_txd_maxcnt, uarte_txd_amount,
+	   uarte_config = 347 };
+/* clang-format on */
+
+static int uartLut[] = { UART0, UART1, UART2, UART3 };
+
+
+static const struct {
+	void *base;
```

Comment:
How does that even work, I wonder, `void *` dereferencing should result in a compilation error
```suggestion
	volatile unsigned int *base;
```